db/migrate/20150716133712_add_unique_index_on_journals.rb

```diff
@@ -0,0 +1,171 @@
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class AddUniqueIndexOnJournals < ActiveRecord::Migration
+  def up
+    cleanup_duplicate_journals
+    add_index :journals, [:journable_type, :journable_id, :version], unique: true
+  end
+
+  def down
+    remove_index :journals, [:journable_type, :journable_id, :version]
+  end
+
+  private
+
+  def cleanup_duplicate_journals
+    duplicate_pairs = find_duplicate_journal_ids
+
+    if duplicate_pairs.any?
+      say "Found #{duplicate_pairs.count} journals with at least one duplicate!"
+      say_with_time 'Safely removing duplicates...' do
+        undeleted_pairs = []
+        duplicate_pairs.each do |current_id, duplicate_id|
+          say "Comparing journals ##{current_id} & ##{duplicate_id} for equality", subitem: true
+
+          current = MigrationHelperJournal.find(current_id)
+          duplicate = MigrationHelperJournal.find(duplicate_id)
+
+          if journals_equivalent?(current, duplicate)
+            say "Deleting journal ##{current.id}...", subitem: true
+            current.destroy
+          else
+            undeleted_pairs << [current_id, duplicate_id]
+          end
```

Comment:
Hm, it might be nicer if we check for the undeleatable_pairs first and abort if there are any and only if non are undeleatable, actually delete all of them. Otherwise the work will be half done.